### DISCUSSION TITLE
From  Mark Novak

### DISCUSSION THREAD

**Discussion by densen2014** [BootstrapBlazor/BootstrapBlazor.Geolocation#2] [Q&A]
Context: [geolocation.js](https://github.com/densen2014/BootstrapBlazor.Geolocation/blob/master/src/BootstrapBlazor.Geolocation/wwwroot/lib/geolocation/geolocation.js) and [Geolocations.razor.cs](https://github.com/densen2014/BootstrapBlazor.Geolocation/blob/master/src/BootstrapBlazor.Geolocation/Geolocations.razor.cs)

In file 1, function line 62 updateLocation at the end  of the function, it invokes a method in the file 2 called GetResult and passes the object of Geolocationitem onto the method. It tries to execute awaitable Func with Invoke.

Question 1:

What possible exception is it trying to catch?

 

Question 2: why does the sample code provided in the README.md return a Task.CompletedTask to OnResult?

Question 2a: is Task.CompletedTask return necessary?

 

Question 3: what was the idea behind implementing returning an object in such way?

**Comment by densen2014** [comment]
A1: maybe is exception from OnResult.Invoke

A2: because is a delegate `Func<Geolocationitem, Task>? OnResult` from geolocations components. may be you need do something async or else in your page.

A3: Delegate `Func<Geolocationitem, Task>`  it is a better way for me , more elegant to write and easier to expand.

```

 [Parameter]
 public Func<Geolocationitem, Task>? OnResult { get; set; }


    /// <summary>
    /// 定位完成回调方法 | Location completion callback method
    /// </summary>
    /// <param name="geolocations"></param>
    /// <returns></returns>
    [JSInvokable]
    public async Task GetResult(Geolocationitem geolocations)
    {
        try
        {
            if (OnResult != null) await OnResult.Invoke(geolocations);
        }
        catch (Exception e)
        {
            if (OnError != null) await OnError.Invoke(e.Message);
        }
    }


```

demo in  https://blazor.app1.es/geolocations

source code 

https://github.com/densen2014/Densen.Extensions/blob/master/Demo/DemoShared/Pages/GeolocationPage.razor

https://github.com/densen2014/Densen.Extensions/blob/master/Demo/DemoShared/Pages/GeolocationPage.razor.cs


```
<Geolocations OnResult="@OnResult" OnUpdateStatus="@OnUpdateStatus" OnError="@OnError" />

<Badge Color="Color.Warning">@status</Badge>

@code{
     private Geolocationitem? geolocations { get; set; }
    private List<Geolocationitem> Items { get; set; } = new List<Geolocationitem>() { new Geolocationitem() };
  
    private Task OnResult(Geolocationitem geolocations)
    {
        this.geolocations = geolocations;
        Items[0] = geolocations;
        StateHasChanged();
        return Task.CompletedTask;
    }
}
```

so

**Reply by tavanuka** [reply]
Thanks a lot! Clears it up. 

- One more question: why does the OnCallbackEvent implementation in the code behind Razor component save it first to an object field, and then only to the first index? 

Is that a performance improvement?

**Comment by densen2014** [comment]
Because in another project of mine, the coordinates are constantly being tracked, so it is a collection. I need to connect into a track ..Densen informatica在 2023年2月10日，21:50，Mark Novak ***@***.***> 写道：﻿
Thanks a lot! Clears it up.

One more question: why does the OnCallbackEvent implementation in the code behind Razor component save it first to an object field, and then only to the first index?

Is that a performance improvement?
